src/Parsers/tests/gtest_format_hiliting.cpp

```diff
@@ -0,0 +1,303 @@
+        left_it++;
+        right_it++;
+    }
+}
+
+TEST(FormatHiliting, MetaTestConsumeHilites)
+{
+    using namespace DB;
+    std::stringstream ss;
+    // The order is different from the order in HILITES on purpose.
+    ss << IAST::hilite_keyword
+       << IAST::hilite_alias
+       << IAST::hilite_identifier
+       << IAST::hilite_none
+       << IAST::hilite_operator
+       << IAST::hilite_substitution
+       << IAST::hilite_function
+       << "test" << IAST::hilite_keyword;
+    std::string string = ss.str();
+    const char * it = string.c_str();
+    const char * expected_it = strchr(it, 't');
+    const char * last_hilite = consume_hilites(&it);
+    ASSERT_EQ(expected_it, it);
+    ASSERT_TRUE(last_hilite != nullptr);
+    ASSERT_EQ(IAST::hilite_function, last_hilite);
+}
+
+TEST(FormatHiliting, MetaTestRemoveHilites)
+{
+    using namespace DB;
+    std::stringstream ss;
+    ss << IAST::hilite_keyword
+       << "te" << IAST::hilite_alias << IAST::hilite_identifier
+       << "s" << IAST::hilite_none
+       << "t" << IAST::hilite_operator << IAST::hilite_substitution << IAST::hilite_function;
+    ASSERT_EQ("test", remove_hilites(ss.str()));
+}
+
+TEST(FormatHiliting, MetaTestAreEqualWithHilites)
+{
+    ASSERT_PRED2(are_equal_with_hilites, "", "");
+
+    for (const char * hilite : HILITES)
+    {
+        ASSERT_PRED2(are_equal_with_hilites, "", std::string_view(hilite));
+        ASSERT_PRED2(are_equal_with_hilites, std::string_view(hilite), "");
+    }
+
+    {
+        std::stringstream ss;
+        ss << DB::IAST::hilite_none << "select" << DB::IAST::hilite_none;
+        ASSERT_PRED2(are_equal_with_hilites, ss.str(), "select");
+    }
+
+    {
+        std::stringstream ss;
+        ss << DB::IAST::hilite_none << "\n " << "sel" << DB::IAST::hilite_none << "ect" << DB::IAST::hilite_none;
+        ASSERT_PRED2(are_equal_with_hilites, ss.str(), "\n select");
+    }
+
+    {
+        std::stringstream left;
+        left << DB::IAST::hilite_keyword << "keyword" << " long" << DB::IAST::hilite_none;
+        std::stringstream right;
+        right << DB::IAST::hilite_keyword << "keyword" << DB::IAST::hilite_none << " " << DB::IAST::hilite_keyword << "long";
+        ASSERT_PRED2(are_equal_with_hilites, left.str(), right.str());
+    }
+}
+
+void compare(const std::string & query, const std::stringstream & expected)
+{
+    using namespace DB;
+    ParserQuery parser(query.data() + query.size());
+    ASTPtr ast = parseQuery(parser, query, 0, 0);
+
+    WriteBufferFromOwnString write_buffer;
+    IAST::FormatSettings settings{write_buffer, true};
+    settings.hilite = true;
+    ast->format(settings);
+
+    ASSERT_PRED2(are_equal_with_hilites, expected.str(), write_buffer.str());
+}
+
+TEST(FormatHiliting, SimpleSelect)
+{
+    std::string query = "select * from table";
+
+    std::stringstream expected;
+    expected << keyword("SELECT ") << "* " << keyword("FROM ") << identifier("table");
```

Comment:
Sure that the space after `SELECT` should be highlighted as well? (it will make no visual difference but still looks wrong)

Reply:
I won't be fixing this, as it is here for user convenience, and it is how it's used everywhere in `formatImpl` overrides. There are hundreds of such cases, and adding space as a separate token in each of them would be inconvenient.

Reply:
Okay, didn't know the funny formatting would be so prevalent 😄 Fine with me to keep things as is. I just would have expected that AstSELECT's `formatImpl()` (conceptually) only hilites `SELECT` and that the caller (... whichever AST type that is - maybe AstQuery but I don't have much clue here) adds the spaces to the output.